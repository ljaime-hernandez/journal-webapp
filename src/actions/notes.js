import Swal from "sweetalert2";
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";


export const startNewNote = () => {
    return async(dispatch, getState) => {

        // retrieves the user id from the store
        const uid = getState().auth.uid;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }
        // asynchronous function pushing a new note into the users journal
        const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

        // we both assign the new note as active on our notesReducer and set it
        // ready on the notesScreen component for us to input the information
        dispatch(activeNote(doc.id, newNote));
        dispatch(addNewNote(doc.id, newNote));
    }
}

// sets the note into an active state, used in the NoteScreen component
export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id, 
        ...note
    }
})

export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id, 
        ...note
    }
})

// uses the loadNotes helper to retrieve an array of notes from the 
// users collection for it to be rendered on the sidebar
export const startLoadingNotes = (uid) => {
    return async(dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));

    } 
}

// sends an array of notes to the notesReducer for them to be loaded
export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote = (note) => {
    return async(dispatch, getState) => {

        // to save the note we retrieve the uid from the actual state on the
        // user store, as some users will forget or wont include a picture
        // on the note, then we make a small condition eliminating the
        // url value from the note
        const { uid } = getState().auth;

        if(!note.url){
            delete note.url;
        }

        // we declare the new note saving on a new variable, then proceeding for us to
        // delete the id, this is done so we reuse the previous note id to update it as
        // the note exists
        const noteToFirestore = {...note};
        delete noteToFirestore.id;

        // we make the request to the firebase db using the users uid, the note id and updating
        // all the items on its document with the update method
        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore);

        // the refresh note will trigger an update on all the existing notes so they are
        // arranged accordingly and they also display the appropriate information on the screen,
        // afterwards sending us a screen popup with Swal to confirm the note was succesfully saved
        dispatch(refreshNote(note.id, noteToFirestore));
        Swal.fire('Saved', note.title, 'success');
    } 
}

export const refreshNote = (id, note) => ({

    type: types.notesUpdate,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const fileUploading = (file) => {
    return async (dispatch, getState) => {
        
        // the fileUploading action can be called only on an active note, either it is
        // a new note or an existing one, so we use the getState method to retrieve
        // the active note information from the store
        const {active:activeNote} = getState().notes;

        // a quick pop up will show using Swal with a loading spinner, while the
        // file is being uploaded using the fileUpload helper which will return the URL
        // where the file is uploaded, followed we call the startSaveNote action which
        // will save the whole note to our firebase collection and finally the
        // pop up will close after the process is done
        Swal.fire({
            title: 'Uploading... please wait',
            text: 'Please wait...',
            allowOutsideClick: false,
            willOpen: () => {
                Swal.showLoading();
            }
        });
        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote))

        Swal.close();
    }
}

export const startDelete = (id) => {
    return async(dispatch, getState) => {

        // similar to our previous actions, we use the getState method to retrieve the
        // user uid, then we use both the note id and the user uid to access to its
        // collection in firebase and trigger the delete method, deleting the document
        // stored on it
        const uid =  getState().auth.uid;

        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id));
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})

export const noteLogout = () => {
    
    return {
        type: types.notesLogoutClean
    }
}