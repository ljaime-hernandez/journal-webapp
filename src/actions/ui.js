import { types } from "../types/types";

export const setError = (error) => {
    return {
        type: types.uiSetError,
        payload: error
    }
}

export const removeError = () => {
    return {
        type: types.uiRemoveError,
    }
}

export const startLoading = () => {
    return {
        type :types.uiStartLoading
    }
}

export const finishLoading = () => {
    return {
        type :types.uiFinishLoading
    }
}