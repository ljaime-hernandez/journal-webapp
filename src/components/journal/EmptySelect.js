import React from 'react';

export const EmptySelect = () => {
  return (
    <div className="empty__main-content">
        <p>
            Select a Journal
            <br/>
            Create an entry
        </p>

        <i className="far fa-star fa-4x mt-5"></i>
    </div>
  )
};
