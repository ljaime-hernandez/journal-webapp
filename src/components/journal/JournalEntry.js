import React from 'react';

export const JournalEntry = () => {
  return (
    <div className="journal__entry pointer">
        <div 
            className="journal__entry-picture"
            style={{
                backgroundSize: 'cover',
                backgroundImage: 'url(https://cms-assets.tutsplus.com/cdn-cgi/image/width=850/uploads/users/523/posts/32694/final_image/tutorial-preview-large.png)'
            }}
        >
        </div>
    
        <div className="journal__entry-body">
            <p className="journal__entry-title">
                New Day
            </p>
            <p className="journal__entry-content">
                ASDFASDFSDASFGSAdffgsaSDFGHDSWDF
                ASDFASDFASDFASDFASDFASDFSADFSADFASDFAS
                DFASDFASDFASDFASDFASDFASDFASDFSDFAS
            </p>
        </div>

        <div className="journal__entry-date-box">
            <span>Monday</span>
            <span>28</span>
        </div>
    </div>
  )
};
