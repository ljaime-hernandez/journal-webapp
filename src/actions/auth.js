import { firebase, googleAuthProvider } from '../firebase/firebaseConfig';
import { startLoading, finishLoading } from './ui';
import { types } from "../types/types";
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

// 
export const startLoginEmailPassword = (email, password) => {

    // the dispatch contains functions from different reducers, but the useDispatch hook
    // is linked to redux, using the environment developed in the store file along with
    // our redux functions, the reducers are contained in the combineReducer method, allowing
    // us to dispatch several actions at once
    return (dispatch) => {

        dispatch(startLoading());

        // account authentication from regular email accounts, it returns a callback which we will
        // use to dispatch the login action from our auth reducer and, either it goes thru or not
        // will also dispatch the finishLoading function to enable the LoginScreen button for login
        firebase.auth().signInWithEmailAndPassword(email, password)
        .then(({user}) => {
            dispatch(finishLoading());
            dispatch(login(user.uid, user.displayName));
        }).catch( e => {
            dispatch(finishLoading());
            Swal.fire('Error', e.message,'error');
        })
    }
}

export const startGoogleLogin = () => {

    // as this function is a callback, the dispatch will be sent back to our authReducer
    // environment for it to handle user login status
    return (dispatch) => {

        // the googleAuthProvider is an authentication method returned from our firebase
        // db, along with the signInWithPopup the user will be able to use its google
        // original account to create an account in our webpage, and for us to
        // use the information for authentication and also render a proper user interface
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName)
                )
            })
        }
}

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}

export const logout = () => {
    return {
        type: types.logout
    }
}

export const startLogout = () => {
    return (dispatch) => {
        firebase.auth().signOut();
        dispatch(noteLogout());
        dispatch(logout());
    }
}
    
export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email, password)
        .then( async ({user}) => {
            await user.updateProfile({displayName: name});

            dispatch(login(user.uid, user.displayName))
        }).catch( e => {
            console.log(e);
            // alert displaying an error message if theres errors on the user input
            // or if the email exists
            Swal.fire('Error', e.message,'error');
        })
    }
}

