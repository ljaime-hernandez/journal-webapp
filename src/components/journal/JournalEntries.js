import React from 'react';
import { useSelector } from 'react-redux';
import { JournalEntry } from './JournalEntry';

export const JournalEntries = () => {

    // uses the information on the store for us to
    // render the existent notes in the users account on 
    // the sidebar, if not, the sidebar will still have a 
    // new entry button for the user to create it.
    const {notes} = useSelector( state => state.notes );

  return (
    <div className="journal__entries animate__animated animate__fadeInLeft">

        {
            notes.map(note => (
                <JournalEntry
                    key={note.id}
                    {...note}
                />
            ))
        }
    </div>
  )
};
