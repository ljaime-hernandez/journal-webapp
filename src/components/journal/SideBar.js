import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';
import { JournalEntries } from './JournalEntries';

export const SideBar = () => {

    // useSelector retrieves the auth object from the store to render it in the
    // sidebar
    const {name} = useSelector( state => state.auth );
    // dispatch will run all our functions to the store using the respective
    // reducer
    const dispatch = useDispatch();
    // the dispatch will use an action which will delete the users
    // information on the reducer
    const handleLogout = () => {
        dispatch(startLogout());
    }
    // the new note will connect to the firebase database, then to the
    // user collection and then will create a document with the respective
    // journal information
    const handleAddNew = () => {
        dispatch(startNewNote());
    }

  return (
    <aside className="journal__sidebar">
        
        <div className="journal__sidebar-navbar">
            <h3 className="mt-5">
                <i className="far fa-moon"/>
                <span>{name}</span>
            </h3>

            <button
                className="btn" 
                onClick={handleLogout}   
            >
                Logout
            </button>
        </div>

        <div
            className="journal__new-entry"    
            onClick={handleAddNew}
        >
            <i className="far fa-calendar-plus fa-5x"></i>
            <p className="mt-5"></p>
            <p>New Entry</p>
            
        </div>

        <JournalEntries/>
    </aside>
  )
};
