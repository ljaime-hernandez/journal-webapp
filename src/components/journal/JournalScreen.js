import React from 'react';
import { SideBar } from './SideBar';

export const JournalScreen = () => {
  return (
    <div className="journal__main-content"> 
        <SideBar/>

        <main>
            <h1>Main content</h1>
        </main>
    </div>)
};
