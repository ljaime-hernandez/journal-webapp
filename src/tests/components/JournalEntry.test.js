import thunk from "redux-thunk";
import configureStore from 'redux-mock-store';
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { JournalEntry } from "../../components/journal/JournalEntry";
import { activeNote } from "../../actions/notes";

/* to run this test:
1. run the 'npm install' command from the journal-webapp folder 
2. run the 'npm install --save-dev enzyme' command (if you have not done so)
3. run the 'npm install --save-dev enzyme-to-json' command (if you have not done so)
4. run the 'npm install --save-dev @wojtekmaj/enzyme-adapter-react-17 --legacy-peer-deps' command (if you are using React 17 as i do)
5. run the 'npm install --save-dev @testing-library/react-hooks' command (if you have not done so)
6. run the 'npm install redux-mock-store --save-dev' command (if you have not done so)
7. make sure the setupTests.js file include the enzyme, enzyme-to-json and the react adapter libraries
8. run the command 'npm run test'
9. to have a clearer view of this single js test file, press p. then type the file name 'JournalEntry.test.js'
*/

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {};

let store = mockStore(initialState);
store.dispatch = jest.fn();

const note = {
    id: 23553,
    date: 10,
    title: 'Test Title',
    body: 'Test Body',
    url: 'https://something.com'
}

describe('tests on JournalEntry component', () => {

    // the journalEntry component should receive all the information in the
    // note object, the note ID will then be used as a reference whenever the
    // entry is clicked, for it to launch further actions in the store and the 
    // active note components
    const wrapper = mount(
        <Provider store={store}>
            <JournalEntry {...note}/>
        </Provider>
    )

    test('should render properly', () => {
       
        expect(wrapper).toMatchSnapshot();
    });

    test('should activate note', () => {
      
        // the div containing the journal__entry class has a click property,
        // which should call the activeNote action and dispatch it to the store,
        // the dispatch does not need to be tested but we need to confirm it was called
        // appropriately with the respective information
        wrapper.find('.journal__entry').prop('onClick')();
        expect(store.dispatch).toHaveBeenCalledWith(
            activeNote(note.id, {...note})
        );
    });
})