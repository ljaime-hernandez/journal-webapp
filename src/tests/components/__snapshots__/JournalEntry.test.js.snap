// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on JournalEntry component should render properly 1`] = `
<div
  className="journal__entry pointer"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://something.com)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Test Title
    </p>
    <p
      className="journal__entry-content"
    >
      Test Body
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      December
    </span>
    <span>
      Wednesday
    </span>
    <span>
      31st
    </span>
    <span>
      7:00 pm
    </span>
  </div>
</div>
`;
